client/client.go

```diff
@@ -0,0 +1,402 @@
+
+// some consts for the jobs returned by NewJob().
+const (
+	jobRetries uint8 = 30
+	reqRAM           = 100
+	reqTime          = 10 * time.Second
+	reqCores         = 1
+	reqDisk          = 1
+)
+
+type SchedulerSettings struct {
+	Deployment  string
+	Cwd         string
+	Queue       string
+	QueuesAvoid string
+	Timeout     time.Duration
+	Logger      log15.Logger
+}
+
+type jobqueueClient interface {
+	Add(jobs []*jobqueue.Job, envVars []string, ignoreComplete bool) (added int, existed int, err error)
+	GetByRepGroup(repgroup string, subStr bool, limit int,
+		state jobqueue.JobState, getStd bool, getEnv bool) ([]*jobqueue.Job, error)
+	Delete(jes []*jobqueue.JobEssence) (int, error)
+	Disconnect() error
+}
+
+type pretendJobqueue struct {
+	jobBuffer []*jobqueue.Job
+	output    io.Writer
+}
+
+func newPretendJobqueue() *pretendJobqueue {
+	var w io.Writer
+
+	fd, errr := strconv.ParseUint(PretendSubmissions, 10, 64)
+	if errr == nil {
+		w = os.NewFile(uintptr(fd), "")
+	}
+
+	return &pretendJobqueue{output: w}
+}
+
+func (p *pretendJobqueue) Add(jobs []*jobqueue.Job, _ []string, _ bool) (int, int, error) {
+	p.jobBuffer = append(p.jobBuffer, jobs...)
+
+	if p.output != nil {
+		json.NewEncoder(p.output).Encode(jobs) //nolint:errcheck,errchkjson
+	}
+
+	return len(jobs), 0, nil
+}
+
+func (p *pretendJobqueue) SubmittedJobs() []*jobqueue.Job {
+	sj := p.jobBuffer
+
+	return sj
+}
+
+// GetByRepGroup behaves like jobqueue.GetByRepGroup, but only repgroup is
+// considered (as a substring).
+func (p *pretendJobqueue) GetByRepGroup(repgroup string, _ bool, _ int,
+	_ jobqueue.JobState, _ bool, _ bool) ([]*jobqueue.Job, error) {
+	var jobs []*jobqueue.Job
+
+	for _, job := range p.jobBuffer {
+		if strings.Contains(job.RepGroup, repgroup) {
+			jobs = append(jobs, job)
+		}
+	}
+
+	return jobs, nil
+}
+
+func (p *pretendJobqueue) Delete(jeses []*jobqueue.JobEssence) (int, error) {
+	origLen := len(p.jobBuffer)
+
+	p.jobBuffer = slices.DeleteFunc(p.jobBuffer, func(job *jobqueue.Job) bool {
+		for _, jes := range jeses {
+			if job.Key() == jes.JobKey {
+				return true
+			}
+		}
+
+		return false
+	})
+
+	return origLen - len(p.jobBuffer), nil
+}
+
+func (p *pretendJobqueue) Disconnect() error {
+	return nil
+}
+
+// Scheduler can be used to schedule commands to be executed by adding them to
+// wr's queue.
+type Scheduler struct {
+	cwd         string
+	exe         string
+	jq          jobqueueClient
+	sudo        bool
+	queue       string
+	queuesAvoid string
+}
+
+// New returns a Scheduler that is connected to wr manager using the given
+// deployment, timeout and logger. Added jobs will have the given cwd, which
+// matters. If cwd is blank, the current working dir is used. If queue is not
+// blank, that queue will be used during NewJob(). If queuesAvoid is not blank,
+// queues including a substring from the list will be avoided during NewJob().
+func New(settings SchedulerSettings) (*Scheduler, error) {
```

Comment:
In a new para, explain again about differing behaviour when PretendSubmissions is set.